misc/connection-limits/src/connection_limits/memory.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use super::*;
+use parking_lot::RwLock;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+/// The configurable memory usage based connection limits.
+#[derive(Debug)]
+pub struct MemoryUsageBasedConnectionLimits {
+    max_process_memory_usage_bytes: Option<usize>,
+    max_process_memory_usage_percentage: Option<f64>,
+    system_physical_memory_bytes: usize,
+    mem_tracker: ProcessMemoryUsageTracker,
+}
+
+impl MemoryUsageBasedConnectionLimits {
+    pub fn new() -> Self {
```

Comment:
Thinking ahead a bit, given that we might have multiple memory-based strategies in the future, I think it makes sense to design the constructor to express which strategies this is. Perhaps something like:

```rust
impl Behaviour {
	/// Configures an absolute memory threshold in bytes.
	///
	/// Once reached, no new connections will be allowed.
	pub fn with_absolute_threshold(max_bytes: usize) -> Self;

	/// Configures a memory threshold in percent, relative to the total physical memory.
	///
	/// Once reached, no more connections will be allowed.
	pub fn with_relative_threshold(percent: u8) -> Self;
}
```

This will also reduce the amount of items we need to expose in the public API. Note that I am assuming here that you never want to use those two options together in one behaviour. Users can still compose this behaviour twice if they want to do that!

Reply:
That makes sense. Maybe it's better to make it an enum instead of struct then

Reply:
For API evolvability reasons, I'd prefer it to be a struct.